### PR TITLE
Add default if empty

### PR DESCRIPTION
**Pull request by sergdort**
Adds defaultIfEmpty operator according to #662

### CODE REVIEW

**Comment by kzaher** [comment]
Hey @sergdort ,

sorry man for not responding, this week was full of setbacks. I'm taking a look right now.

File: RxSwift/Observables/Implementations/DefaultIfEmpty.swift [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+//  DefaultIfEmpty.swift
+//  Rx
+//
+//  Created by sergdort on 23/12/2016.
+//  Copyright © 2016 Krunoslav Zaher. All rights reserved.
+//
+
+import Foundation
+
+class DefaultIfEmptySink<SourceType, O: ObserverType>: Sink<O>, ObserverType where O.E == SourceType {
```
Comment (kzaher) [inline RxSwift/Observables/Implementations/DefaultIfEmpty.swift]: We recently had a refactoring change merged.

https://github.com/ReactiveX/RxSwift/pull/1045

I think @tarunon was right, we could make this simpler.

File: RxSwift/Observables/Implementations/DefaultIfEmpty.swift [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    private let _defaultValue: SourceType
+    private var isEmpty = true
```
Comment (kzaher) [inline RxSwift/Observables/Implementations/DefaultIfEmpty.swift]: _isEmpty :)

I know that we are using maybe a slightly weird convention, but it helps a lot when reasoning about local state of the operator.

File: RxSwift/Observables/Implementations/DefaultIfEmpty.swift [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            dispose()
+        }
+    }
+}
+
+class DefaultIfEmpty<SourceType>: Producer<SourceType> {
+    private let _source: Observable<SourceType>
+    private let _defautValue: SourceType
+    
+    init(source: Observable<SourceType>, defautValue: SourceType) {
```
Comment (kzaher) [inline RxSwift/Observables/Implementations/DefaultIfEmpty.swift]: `defaultValue`

File: RxSwift/Observables/Implementations/DefaultIfEmpty.swift [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            }
+            forwardOn(.completed)
+            dispose()
+        }
+    }
+}
+
+class DefaultIfEmpty<SourceType>: Producer<SourceType> {
+    private let _source: Observable<SourceType>
+    private let _defautValue: SourceType
```
Comment (kzaher) [inline RxSwift/Observables/Implementations/DefaultIfEmpty.swift]: `_defaultValue`

File: RxSwift/Observables/Observable+Single.swift [hunk 5]
```diff
@@ -196,3 +196,17 @@ extension ObservableType {
...
+// MARK: defaultIfEmpty
+
+extension ObservableType {
+    /**
+    Emit items from the source Observable, or a default item if the source Observable emits nothing
+     
+     - seealso: [DefaultIfEmpty operator on reactivex.io](http://reactivex.io/documentation/operators/defaultifempty.html)
+     - parameter defautValue: An default value to be sent if the source does not emit any elements
+    */
+    public func defaultIfEmpty(_ defautValue: E) -> Observable<E> {
```
Comment (kzaher) [inline RxSwift/Observables/Observable+Single.swift]: `defautValue` :)

I'm thinking that the name would probably be more in the spirit of Swift if it was something like.

```
public func ifEmpty(default: E) -> Observable<E>
```

and corresponding

```
public func ifEmpty(switch: Observable<E>) -> Observable<E>
```

File: Tests/RxSwiftTests/Observable+SingleTest.swift [hunk 6]
```diff
@@ -1590,3 +1590,84 @@ extension ObservableSingleTest {
             ])
     }
 }
+
+// defaultIfEmpty
+
+extension ObservableSingleTest {
```
Comment (kzaher) [inline Tests/RxSwiftTests/Observable+SingleTest.swift]: Unit tests look great to me.

File: RxSwift/Observables/Implementations/DefaultIfEmpty.swift (hunk 1 above)
Comment (kzaher) [inline RxSwift/Observables/Implementations/DefaultIfEmpty.swift]: We can also probably add `final class`.

If we don't have it everywhere we can, we should add it.

Comment (sergdort) [inline RxSwift/Observables/Implementations/DefaultIfEmpty.swift]: Yes indeed, from what I heard it decreases build time and compiler does some optimizations apparently

File: RxSwift/Observables/Observable+Single.swift (hunk 5 above)
Comment (sergdort) [inline RxSwift/Observables/Observable+Single.swift]: Cool 👍

**Comment by sergdort** [comment]
Have no clue why it's failing 😞

**Comment by tarunon** [comment]
I think you forgot running test scripts. It's for Linux environment.
`% ./scripts/all-tests.sh`

**Comment by sergdort** [comment]
Thanks @tarunon !

**Comment by kzaher** [comment]
Hi @sergdort ,

I think there are still open issues from the original comments. I'm not sure why github marked them as outdated.

**Comment by sergdort** [comment]
Yeah, damn it GitHub 😁. I believe now it should be fine

File: RxSwift/Observables/Observable+Single.swift [hunk 7]
```diff
@@ -197,7 +197,22 @@ extension ObservableType {
...
+
+extension ObservableType {
+    
+    /**
+     Emit items from the source Observable, or a default item if the source Observable emits nothing
+     
+     - seealso: [DefaultIfEmpty operator on reactivex.io](http://reactivex.io/documentation/operators/defaultifempty.html)
+     
+     - parameter default: An default value to be sent if the source does not emit any elements
+     - returns: An observable sequence wich emits default element end completes in case the original sequance is empty
```
Comment (kzaher) [inline RxSwift/Observables/Observable+Single.swift]: `which`, `sequence`, `and` ... `sequence`

File: RxSwift/Observables/Observable+Single.swift [hunk 8]
```diff
@@ -197,7 +197,22 @@ extension ObservableType {
     }
 }
 
-// MARK: ignoreElements
+// MARK: defaultIfEmpty
+
+extension ObservableType {
+    
+    /**
+     Emit items from the source Observable, or a default item if the source Observable emits nothing
```
Comment (kzaher) [inline RxSwift/Observables/Observable+Single.swift]: `elements` vs `items`, `Observable` sequence.

File: RxSwift/Observables/Observable+Single.swift [hunk 9]
```diff
@@ -197,7 +197,22 @@ extension ObservableType {
...
+     
+     - seealso: [DefaultIfEmpty operator on reactivex.io](http://reactivex.io/documentation/operators/defaultifempty.html)
+     
+     - parameter default: An default value to be sent if the source does not emit any elements
```
Comment (kzaher) [inline RxSwift/Observables/Observable+Single.swift]: `A` vs `An`

**Comment by sergdort** [comment]
Hey, @kzaher , sorry man :) do you whiny me to address  your last comments  ?

**Pull request merged**